src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Ssl.cs

```diff
@@ -279,4 +290,86 @@ private SafeSslHandle() : base(IntPtr.Zero, true)
         {
         }
     }
+
+    internal sealed class SafeChannelBinding : SafeHandle
+    {
+        [StructLayout(LayoutKind.Sequential)]
+        private struct SEC_CHANNEL_BINDINGS
+        {
+            internal int dwInitiatorAddrType;
+            internal int cbInitiatorLength;
+            internal int dwInitiatorOffset;
+            internal int dwAcceptorAddrType;
+            internal int cbAcceptorLength;
+            internal int dwAcceptorOffset;
+            internal int cbApplicationDataLength;
+            internal int dwApplicationDataOffset;
+        }
+
+        private static readonly byte[] s_tlsServerEndPointByteArray = Encoding.UTF8.GetBytes("tls-server-end-point:");
+        private static readonly byte[] s_tlsUniqueByteArray = Encoding.UTF8.GetBytes("tls-unique:");
+        private static readonly int secChannelBindingSize = Marshal.SizeOf<SEC_CHANNEL_BINDINGS>();
+        private readonly int cbtPrefixByteArraySize;
+        private const int CertHashMaxSize = 128;
+
+        internal int Length
+        {
+            get;
+            private set;
+        }
+
+        internal IntPtr CertHashPtr
+        {
+            get;
+            private set;
+        }
+
+        private byte[] GetPrefixBytes(ChannelBindingKind kind)
+        {
+            if (kind == ChannelBindingKind.Endpoint)
+            {
+                return s_tlsServerEndPointByteArray;
+            }
+            else
+            {
+                return s_tlsUniqueByteArray;
+            }
+        }
+
+        internal SafeChannelBinding(ChannelBindingKind kind)
+            : base(IntPtr.Zero, true)
+        {
+            byte[] cbtPrefix = GetPrefixBytes(kind);
+            cbtPrefixByteArraySize = cbtPrefix.Length;
+            handle = Marshal.AllocHGlobal(secChannelBindingSize + cbtPrefixByteArraySize + CertHashMaxSize);
+            IntPtr cbtPrefixPtr = handle + secChannelBindingSize;
+            Marshal.Copy(cbtPrefix, 0, cbtPrefixPtr, cbtPrefixByteArraySize);
+            CertHashPtr = cbtPrefixPtr + cbtPrefixByteArraySize;
+            Length = CertHashMaxSize;
+        }
+
+        internal void SetCertHashLength(int certHashLength)
+        {
+            int cbtLength = cbtPrefixByteArraySize + certHashLength;
+            Length = secChannelBindingSize + cbtLength;
+
+            SEC_CHANNEL_BINDINGS channelBindings = new SEC_CHANNEL_BINDINGS()
+            {
+                cbApplicationDataLength = cbtLength,
+                dwApplicationDataOffset = secChannelBindingSize
+            };
+            Marshal.StructureToPtr(channelBindings, handle, true);
+        }
+
+        public override bool IsInvalid
+        {
+            get { return handle == IntPtr.Zero; }
+        }
+
+        protected override bool ReleaseHandle()
+        {
+            Marshal.FreeHGlobal(handle);
+            return true;
```

Comment:
SetHandle(IntPtr.Zero) is a good idea, so that IsInvalid returns true after this point.  (Otherwise you have to remember to check IsClosed)